File: src/lib/api/realtime.ts
import { Channel, invoke } from "@tauri-apps/api/core";

import { hasTauriRuntime } from "$lib/api/commands";
import type {
  ExportResult,
  RealtimeConnectionDraft,
  RealtimeConnectionProfileDetail,
  RealtimeConnectionProfileSummary,
  RealtimeMessageDraft,
  RealtimeRuntimeEvent,
  RealtimeSessionSnapshot,
  RealtimeWorkspaceState,
  SavedRealtimeMessageDetail,
  SavedRealtimeMessageSummary
} from "$lib/api/types";
import { createRealtimeConnectionDraft, createRealtimeMessageDraft } from "$lib/api/types";

export type RealtimeConnectInput = { sessionId: string; connection: RealtimeConnectionDraft };
export type RealtimeSendMessage = RealtimeMessageDraft;
export type RealtimeEventSubscription = { close: () => void };

const mockSavedMessages = new Map<string, SavedRealtimeMessageDetail>();
const mockProfiles = new Map<string, RealtimeConnectionProfileDetail>();
let mockWorkspaceState: RealtimeWorkspaceState | null = null;
const mockSessions = new Map<string, { snapshot: RealtimeSessionSnapshot; protocol: RealtimeConnectionDraft["protocol"]; onEvent: (event: RealtimeRuntimeEvent) => void }>();

const now = () => new Date().toISOString();
const inlinePayload = (text: string, encoding: "utf8" | "base64" = "utf8") => ({
  mode: "inline" as const, text, sizeBytes: new TextEncoder().encode(text).byteLength, encoding, truncated: false
});

function createMockSnapshot(sessionId: string): RealtimeSessionSnapshot {
  return { sessionId, generation: 1, lastSequence: 1, status: "connected", statusMessage: "Connected", transcript: [{
    id: `${sessionId}-1`, sessionId, generation: 1, sequence: 1, occurredAt: now(), direction: "system",
    kind: "lifecycle", label: "Connected", eventName: null, payload: null
  }], transcriptSizeBytes: 0 };
}

function cloneSummary(detail: SavedRealtimeMessageDetail): SavedRealtimeMessageSummary {
  const { message: _message, ...summary } = detail;
  return structuredClone(summary);
}

function seedMessage(itemId: string): SavedRealtimeMessageDetail | null {
  const protocol = itemId === "mock-realtime-socketio-1" ? "socketio" : itemId === "mock-realtime-websocket-1" ? "websocket" : null;
  if (!protocol) return null;
  const message = createRealtimeMessageDraft(protocol);
  message.name = protocol === "socketio" ? "Support presence" : "Live order events";
  return { id: itemId, collectionId: "mock-collection-1", parentId: null, name: message.name, requestType: protocol, updatedAt: "2026-07-30T12:00:00.000Z", message };
}

function seedProfiles() {
  if (mockProfiles.size) return;
  for (const protocol of ["websocket", "socketio"] as const) {
    const connection = createRealtimeConnectionDraft(protocol);
    connection.name = protocol === "websocket" ? "Order events" : "Support presence";
    connection.url = protocol === "websocket" ? "wss://events.example.test/orders" : "https://presence.example.test";
    const id = `mock-${protocol}-profile`;
    mockProfiles.set(id, { id, name: connection.name, protocol, url: connection.url, updatedAt: now(), connection });
  }
}

export async function getRealtimeWorkspaceState() {
  if (!hasTauriRuntime()) return mockWorkspaceState;
  return invoke<RealtimeWorkspaceState | null>("get_realtime_workspace_state");
}
export async function saveRealtimeWorkspaceState(state: RealtimeWorkspaceState) {
  if (!hasTauriRuntime()) { mockWorkspaceState = structuredClone(state); return; }
  await invoke("save_realtime_workspace_state", { state });
}

export async function listRealtimeConnectionProfiles(): Promise<RealtimeConnectionProfileSummary[]> {
  if (!hasTauriRuntime()) { seedProfiles(); return [...mockProfiles.values()].map(({ connection: _, ...item }) => structuredClone(item)); }
  return invoke("list_realtime_connection_profiles");
}
export async function getRealtimeConnectionProfile(profileId: string): Promise<RealtimeConnectionProfileDetail> {
  if (!hasTauriRuntime()) { seedProfiles(); const item = mockProfiles.get(profileId); if (item) return structuredClone(item); throw new Error("Realtime connection profile not found."); }
  return invoke("get_realtime_connection_profile", { profileId });
}
export async function createRealtimeConnectionProfile(connection: RealtimeConnectionDraft): Promise<RealtimeConnectionProfileDetail> {
  if (!hasTauriRuntime()) { const detail = { id: `mock-profile-${Date.now()}`, name: connection.name, protocol: connection.protocol, url: connection.url, updatedAt: now(), connection: structuredClone(connection) }; mockProfiles.set(detail.id, detail); return detail; }
  return invoke("create_realtime_connection_profile", { connection });
}
export async function updateRealtimeConnectionProfile(profileId: string, connection: RealtimeConnectionDraft, expectedUpdatedAt?: string | null): Promise<RealtimeConnectionProfileDetail> {
  if (!hasTauriRuntime()) { const detail = { id: profileId, name: connection.name, protocol: connection.protocol, url: connection.url, updatedAt: now(), connection: structuredClone(connection) }; mockProfiles.set(profileId, detail); return detail; }
  return invoke("update_realtime_connection_profile", { profileId, connection, expectedUpdatedAt: expectedUpdatedAt ?? null });
}
export async function deleteRealtimeConnectionProfile(profileId: string, expectedUpdatedAt?: string | null): Promise<void> {
  if (!hasTauriRuntime()) { mockProfiles.delete(profileId); return; }
  await invoke("delete_realtime_connection_profile", { profileId, expectedUpdatedAt: expectedUpdatedAt ?? null });
}
export async function importRealtimeConnectionProfiles(): Promise<RealtimeConnectionProfileDetail[]> {
  if (!hasTauriRuntime()) return [];
  return invoke("import_realtime_connection_profiles");
}
export async function exportRealtimeConnectionProfiles(profileIds: string[], includeSensitive = false): Promise<ExportResult | null> {
  if (!hasTauriRuntime()) return { filePath: "/tmp/postnot-realtime-connections.json" };
  return invoke("export_realtime_connection_profiles", { profileIds, includeSensitive });
}

export async function saveRealtimeMessageToCollection(collectionId: string, message: RealtimeMessageDraft, parentId?: string | null): Promise<SavedRealtimeMessageSummary> {
  if (!hasTauriRuntime()) { const detail: SavedRealtimeMessageDetail = { id: `mock-realtime-message-${Date.now()}`, collectionId, parentId: parentId ?? null, name: message.name, requestType: message.protocol, updatedAt: now(), message: structuredClone(message) }; mockSavedMessages.set(detail.id, detail); return cloneSummary(detail); }
  return invoke("save_realtime_message_to_collection", { collectionId, parentId: parentId ?? null, message });
}
export async function updateSavedRealtimeMessage(itemId: string, message: RealtimeMessageDraft, expectedUpdatedAt?: string | null): Promise<SavedRealtimeMessageSummary> {
  if (!hasTauriRuntime()) { const old = mockSavedMessages.get(itemId); const detail: SavedRealtimeMessageDetail = { id: itemId, collectionId: old?.collectionId ?? "mock-collection-1", parentId: old?.parentId ?? null, name: message.name, requestType: message.protocol, updatedAt: now(), message: structuredClone(message) }; mockSavedMessages.set(itemId, detail); return cloneSummary(detail); }
  return invoke("update_saved_realtime_message", { itemId, message, expectedUpdatedAt: expectedUpdatedAt ?? null });
}
export async function getSavedRealtimeMessage(itemId: string): Promise<SavedRealtimeMessageDetail> {
  if (!hasTauriRuntime()) { const item = mockSavedMessages.get(itemId) ?? seedMessage(itemId); if (item) return structuredClone(item); throw new Error("Saved realtime message not found."); }
  return invoke("get_saved_realtime_message", { itemId });
}
export async function listSavedRealtimeMessages(collectionId: string): Promise<SavedRealtimeMessageSummary[]> {
  if (!hasTauriRuntime()) return [seedMessage("mock-realtime-websocket-1"), seedMessage("mock-realtime-socketio-1"), ...mockSavedMessages.values()].filter((item): item is SavedRealtimeMessageDetail => Boolean(item && item.collectionId === collectionId)).map(cloneSummary);
  return invoke("list_saved_realtime_messages", { collectionId });
}
export async function deleteSavedRealtimeMessage(itemId: string): Promise<void> {
  if (!hasTauriRuntime()) { mockSavedMessages.delete(itemId); return; }
  await invoke("delete_saved_realtime_message", { itemId });
}

function emitMockTranscript(sessionId: string, entry: Omit<RealtimeSessionSnapshot["transcript"][number], "id" | "sessionId" | "generation" | "sequence" | "occurredAt">) {
  const session = mockSessions.get(sessionId); if (!session) throw new Error("Connect before sending a message.");
  const sequence = ++session.snapshot.lastSequence;
  const complete = { ...entry, id: `${sessionId}-${sequence}`, sessionId, generation: session.snapshot.generation, sequence, occurredAt: now() };
  session.snapshot.transcript.push(complete); session.snapshot.transcriptSizeBytes += complete.payload?.sizeBytes ?? 0;
  session.onEvent({ type: "transcript", sessionId, generation: session.snapshot.generation, sequence, entry: structuredClone(complete) });
}

function mockPresentation(message: RealtimeMessageDraft) {
  if (message.protocol === "websocket") {
    if (message.composer.mode === "binary") { const binary = message.composer.binary; return { kind: "binary" as const, label: "Binary message", eventName: null, payload: inlinePayload(binary?.source === "file" ? binary.path : binary?.value ?? "", "base64") }; }
    return { kind: message.composer.mode, label: message.composer.mode === "json" ? "JSON message" : "Text message", eventName: null, payload: inlinePayload(message.composer.content) };
  }
  const text = message.composer.binary ? (message.composer.binary.source === "file" ? message.composer.binary.path : message.composer.binary.value) : JSON.stringify(message.composer.arguments);
  return { kind: message.composer.binary ? "binary" as const : "event" as const, label: message.composer.waitForAck ? "Event · awaiting ACK" : "Event", eventName: message.composer.event, payload: inlinePayload(text, message.composer.binary ? "base64" : "utf8") };
}

export async function connectRealtimeConnection(input: RealtimeConnectInput, onEvent: (event: RealtimeRuntimeEvent) => void): Promise<{ result: RealtimeSessionSnapshot; subscription: RealtimeEventSubscription }> {
  if (!hasTauriRuntime()) { const snapshot = createMockSnapshot(input.sessionId); mockSessions.set(input.sessionId, { snapshot, protocol: input.connection.protocol, onEvent }); return { result: structuredClone(snapshot), subscription: { close: () => { const session = mockSessions.get(input.sessionId); if (session) session.onEvent = () => {}; } } }; }
  const channel = new Channel<RealtimeRuntimeEvent>(); channel.onmessage = onEvent;
  const result = await invoke<RealtimeSessionSnapshot>("connect_realtime_connection", { input, onEvent: channel });
  return { result, subscription: { close: () => { channel.onmessage = () => {}; } } };
}
export async function disconnectRealtimeConnection(sessionId: string) { if (!hasTauriRuntime()) { const session = mockSessions.get(sessionId); if (session) { session.snapshot.status = "disconnected"; session.snapshot.statusMessage = "Disconnected"; } return; } await invoke("disconnect_realtime_connection", { sessionId }); }
export async function releaseRealtimeConnection(sessionId: string) { if (!hasTauriRuntime()) { mockSessions.delete(sessionId); return; } await invoke("release_realtime_connection", { sessionId }); }
export async function sendRealtimeMessage(sessionId: string, message: RealtimeMessageDraft) {
  if (!hasTauriRuntime()) { const session = mockSessions.get(sessionId); if (!session) throw new Error("Connect before sending a message."); if (session.protocol !== message.protocol) throw new Error(`Connected ${session.protocol} session cannot send a ${message.protocol} message.`); const p = mockPresentation(message); emitMockTranscript(sessionId, { ...p, direction: "sent" }); emitMockTranscript(sessionId, { ...p, direction: "received", label: message.protocol === "socketio" ? "Mock server event" : "Mock echo" }); return; }
  await invoke("send_realtime_message", { sessionId, message });
}
export async function pingRealtimeConnection(sessionId: string, payload?: string) { if (!hasTauriRuntime()) { emitMockTranscript(sessionId, { direction: "sent", kind: "ping", label: "Ping", eventName: null, payload: payload ? inlinePayload(payload) : null }); return; } await invoke("ping_realtime_connection", { sessionId, payload: payload ?? null }); }
export async function closeRealtimeConnection(sessionId: string, code = 1000, reason = "") { if (!hasTauriRuntime()) { const session = mockSessions.get(sessionId); if (session) { session.snapshot.status = "disconnected"; session.snapshot.statusMessage = "Disconnected"; } return; } await invoke("close_realtime_connection", { sessionId, code, reason }); }
export async function getRealtimeSessionSnapshot(sessionId: string): Promise<RealtimeSessionSnapshot> { if (!hasTauriRuntime()) { const snapshot = mockSessions.get(sessionId)?.snapshot; if (!snapshot) throw new Error("Realtime session not found."); return structuredClone(snapshot); } return invoke("get_realtime_session_snapshot", { sessionId }); }
export async function clearRealtimeTranscript(sessionId: string) { if (!hasTauriRuntime()) { const session = mockSessions.get(sessionId); if (session) { session.snapshot.transcript = []; session.snapshot.transcriptSizeBytes = 0; } return; } await invoke("clear_realtime_transcript", { sessionId }); }
export async function readRealtimePayload(handleId: string): Promise<string> { if (!hasTauriRuntime()) return ""; return invoke("read_realtime_payload", { handleId }); }
export async function saveRealtimePayload(handleId: string, suggestedName?: string): Promise<string | null> { if (!hasTauriRuntime()) return null; return invoke("save_realtime_payload", { handleId, suggestedName: suggestedName ?? null }); }
export async function exportRealtimeTranscript(sessionId: string): Promise<ExportResult | null> { if (!hasTauriRuntime()) return { filePath: `/tmp/${sessionId}-transcript.json` }; return invoke("export_realtime_transcript", { sessionId }); }

File: src-tauri/src/commands/realtime.rs
use std::{
    fs::File,
    io::{BufReader, BufWriter, Read, Write},
    path::Path,
    time::Duration,
};

use base64::{engine::general_purpose::STANDARD as BASE64, write::EncoderWriter};
use tauri::{ipc::Channel, State};

use crate::{
    app_state::AppState,
    domain::exports::ExportResult,
    error::AppResult,
    services::{
        environments_service,
        realtime_payload_service::RealtimePayload,
        realtime_resolution_service,
        realtime_service::{
            RealtimeConnectInput, RealtimeRuntimeEvent, RealtimeRuntimeLimits, RealtimeSendMessage,
            RealtimeSessionSnapshot,
        },
        settings_service,
    },
};

#[tauri::command]
pub async fn get_realtime_workspace_state(
    state: State<'_, AppState>,
) -> AppResult<Option<serde_json::Value>> {
    settings_service::get_realtime_workspace_state(state.db()).await
}

#[tauri::command]
pub async fn save_realtime_workspace_state(
    app_state: State<'_, AppState>,
    state: serde_json::Value,
) -> AppResult<()> {
    settings_service::save_realtime_workspace_state(app_state.db(), &state).await
}

#[tauri::command]
pub async fn connect_realtime_connection(
    state: State<'_, AppState>,
    input: RealtimeConnectInput,
    on_event: Channel<RealtimeRuntimeEvent>,
) -> AppResult<RealtimeSessionSnapshot> {
    let active_environment =
        environments_service::get_active_environment(state.db(), state.secret_store()).await?;
    let resolved =
        realtime_resolution_service::resolve_request(&input.request, active_environment.as_ref());
    let settings = settings_service::get_settings(state.db()).await?;
    let limits = RealtimeRuntimeLimits {
        connect_timeout: Duration::from_millis(settings.realtime_connect_timeout_ms),
        max_concurrent_sessions: settings.realtime_max_concurrent_sessions as usize,
        max_message_bytes: usize::try_from(settings.realtime_max_message_bytes)
            .unwrap_or(usize::MAX),
        transcript_max_entries: settings.realtime_transcript_max_entries as usize,
        transcript_max_bytes: settings.realtime_transcript_max_bytes,
        validate_tls: settings.validate_tls,
    };
    state
        .realtime_connections()
        .connect(
            input,
            resolved.request,
            resolved.secret_values,
            limits,
            on_event,
        )
        .await
}

#[tauri::command]
pub async fn disconnect_realtime_connection(
    state: State<'_, AppState>,
    connection_id: String,
) -> AppResult<()> {
    state
        .realtime_connections()
        .disconnect(&connection_id)
        .await
}

#[tauri::command]
pub async fn release_realtime_connection(
    state: State<'_, AppState>,
    connection_id: String,
) -> AppResult<()> {
    state.realtime_connections().release(&connection_id).await
}

#[tauri::command]
pub async fn send_realtime_message(
    state: State<'_, AppState>,
    connection_id: String,
    message: RealtimeSendMessage,
) -> AppResult<()> {
    let active_environment =
        environments_service::get_active_environment(state.db(), state.secret_store()).await?;
    let secret_values =
        environments_service::active_environment_secret_values(active_environment.as_ref());
    let message = match message {
        RealtimeSendMessage::Websocket { composer } => RealtimeSendMessage::Websocket {
            composer: realtime_resolution_service::resolve_raw_composer(
                &composer,
                active_environment.as_ref(),
            ),
        },
        RealtimeSendMessage::Socketio { composer } => RealtimeSendMessage::Socketio {
            composer: realtime_resolution_service::resolve_socketio_composer(
                &composer,
                active_environment.as_ref(),
            ),
        },
    };
    state
        .realtime_connections()
        .send(&connection_id, message, secret_values)
        .await
}

#[tauri::command]
pub async fn ping_realtime_connection(
    state: State<'_, AppState>,
    connection_id: String,
    payload: Option<String>,
) -> AppResult<()> {
    state
        .realtime_connections()
        .ping(&connection_id, payload)
        .await
}

#[tauri::command]
pub async fn close_realtime_connection(
    state: State<'_, AppState>,
    connection_id: String,
    code: u16,
    reason: String,
) -> AppResult<()> {
    state
        .realtime_connections()
        .close(&connection_id, code, reason)
        .await
}

#[tauri::command]
pub fn get_realtime_session_snapshot(
    state: State<'_, AppState>,
    connection_id: String,
) -> AppResult<RealtimeSessionSnapshot> {
    state.realtime_connections().snapshot(&connection_id)
}

#[tauri::command]
pub async fn clear_realtime_transcript(
    state: State<'_, AppState>,
    connection_id: String,
) -> AppResult<()> {
    state
        .realtime_connections()
        .clear_transcript(&connection_id)
        .await
}

#[tauri::command]
pub async fn read_realtime_payload(
    state: State<'_, AppState>,
    handle_id: String,
) -> AppResult<String> {
    state
        .realtime_connections()
        .payloads()
        .read(&handle_id)
        .await
}

#[tauri::command]
pub async fn save_realtime_payload(
    state: State<'_, AppState>,
    handle_id: String,
    suggested_name: Option<String>,
) -> AppResult<Option<String>> {
    let suggested_name = suggested_name
        .filter(|name| !name.trim().is_empty())
        .map(|name| sanitize_suggested_file_name(&name))
        .unwrap_or_else(|| "realtime-payload.bin".to_string());
    let path = tauri::async_runtime::spawn_blocking(move || {
        rfd::FileDialog::new()
            .set_title("Save realtime payload")
            .set_file_name(&suggested_name)
            .save_file()
    })
    .await?;
    let Some(path) = path else {
        return Ok(None);
    };
    state
        .realtime_connections()
        .payloads()
        .copy_to(&handle_id, &path)
        .await?;
    Ok(Some(path.to_string_lossy().to_string()))
}

#[tauri::command]
pub async fn export_realtime_transcript(
    state: State<'_, AppState>,
    connection_id: String,
) -> AppResult<Option<ExportResult>> {
    let (snapshot, retained_handles) = state
        .realtime_connections()
        .snapshot_for_export(&connection_id)?;
    let suggested_name = format!(
        "{}-transcript.json",
        sanitize_file_stem(&connection_id, "realtime")
    );
    let path = match choose_transcript_path(suggested_name).await {
        Ok(path) => path,
        Err(error) => {
            for handle in retained_handles {
                let _ = state
                    .realtime_connections()
                    .payloads()
                    .release(&handle)
                    .await;
            }
            return Err(error);
        }
    };
    let Some(path) = path else {
        for handle in retained_handles {
            state
                .realtime_connections()
                .payloads()
                .release(&handle)
                .await?;
        }
        return Ok(None);
    };
    let payloads = state.realtime_connections().payloads().clone();
    let export_path = path.clone();
    let export_result = tauri::async_runtime::spawn_blocking(move || {
        write_transcript_export(&export_path, &snapshot, &payloads)
    })
    .await;
    for handle in retained_handles {
        state
            .realtime_connections()
            .payloads()
            .release(&handle)
            .await?;
    }
    export_result??;
    Ok(Some(ExportResult {
        file_path: path.to_string_lossy().to_string(),
    }))
}

async fn choose_transcript_path(suggested_name: String) -> AppResult<Option<std::path::PathBuf>> {
    let path = tauri::async_runtime::spawn_blocking(move || {
        rfd::FileDialog::new()
            .set_title("Export realtime transcript")
            .set_file_name(&suggested_name)
            .add_filter("JSON", &["json"])
            .save_file()
    })
    .await?;
    Ok(path)
}

fn sanitize_file_stem(input: &str, fallback: &str) -> String {
    let sanitized = input
        .chars()
        .map(|character| {
            if character.is_ascii_alphanumeric() || matches!(character, '-' | '_') {
                character
            } else {
                '-'
            }
        })
        .collect::<String>()
        .trim_matches('-')
        .to_string();
    if sanitized.is_empty() {
        fallback.to_string()
    } else {
        sanitized
    }
}

fn sanitize_suggested_file_name(input: &str) -> String {
    let file_name = input
        .rsplit(['/', '\\'])
        .next()
        .unwrap_or("realtime-payload.bin");
    let (stem, extension) = file_name
        .rsplit_once('.')
        .filter(|(_, extension)| {
            !extension.is_empty()
                && extension.len() <= 16
                && extension
                    .chars()
                    .all(|character| character.is_ascii_alphanumeric())
        })
        .unwrap_or((file_name, "bin"));
    format!(
        "{}.{}",
        sanitize_file_stem(stem, "realtime-payload"),
        extension
    )
}

fn write_transcript_export(
    path: &Path,
    snapshot: &RealtimeSessionSnapshot,
    payloads: &crate::services::realtime_payload_service::RealtimePayloadStore,
) -> AppResult<()> {
    let mut writer = BufWriter::new(File::create(path)?);
    write!(writer, "{{\"connectionId\":")?;
    serde_json::to_writer(&mut writer, &snapshot.connection_id)?;
    write!(
        writer,
        ",\"generation\":{},\"lastSequence\":{},\"status\":",
        snapshot.generation, snapshot.last_sequence
    )?;
    serde_json::to_writer(&mut writer, &snapshot.status)?;
    write!(writer, ",\"statusMessage\":")?;
    serde_json::to_writer(&mut writer, &snapshot.status_message)?;
    write!(
        writer,
        ",\"transcriptSizeBytes\":{},\"transcript\":[",
        snapshot.transcript_size_bytes
    )?;
    for (index, entry) in snapshot.transcript.iter().enumerate() {
        if index > 0 {
            writer.write_all(b",")?;
        }
        match entry.payload.as_ref() {
            Some(RealtimePayload::File {
                handle_id,
                size_bytes,
                encoding,
                ..
            }) => {
                let (source, stored_encoding) = payloads.export_source(handle_id)?;
                write_export_entry_prefix(&mut writer, entry)?;
                write!(writer, "{{\"mode\":\"inline\",\"text\":")?;
                match stored_encoding {
                    crate::services::realtime_payload_service::RealtimePayloadEncoding::Utf8 => {
                        write_json_string_file(&mut writer, &source)?;
                    }
                    crate::services::realtime_payload_service::RealtimePayloadEncoding::Base64 => {
                        writer.write_all(b"\"")?;
                        let mut source = BufReader::new(File::open(source)?);
                        {
                            let mut encoder = EncoderWriter::new(&mut writer, &BASE64);
                            std::io::copy(&mut source, &mut encoder)?;
                            let _ = encoder.finish()?;
                        }
                        writer.write_all(b"\"")?;
                    }
                }
                write!(writer, ",\"sizeBytes\":{},\"encoding\":", size_bytes)?;
                serde_json::to_writer(&mut writer, encoding)?;
                writer.write_all(b",\"truncated\":false}}")?;
            }
            _ => serde_json::to_writer(&mut writer, entry)?,
        }
    }
    writer.write_all(b"]}")?;
    writer.flush()?;
    Ok(())
}

fn write_export_entry_prefix(
    writer: &mut impl Write,
    entry: &crate::services::realtime_service::RealtimeTranscriptEntry,
) -> AppResult<()> {
    writer.write_all(b"{\"id\":")?;
    serde_json::to_writer(&mut *writer, &entry.id)?;
    writer.write_all(b",\"connectionId\":")?;
    serde_json::to_writer(&mut *writer, &entry.connection_id)?;
    write!(
        writer,
        ",\"generation\":{},\"sequence\":{},\"occurredAt\":",
        entry.generation, entry.sequence
    )?;
    serde_json::to_writer(&mut *writer, &entry.occurred_at)?;
    writer.write_all(b",\"direction\":")?;
    serde_json::to_writer(&mut *writer, &entry.direction)?;
    writer.write_all(b",\"kind\":")?;
    serde_json::to_writer(&mut *writer, &entry.kind)?;
    writer.write_all(b",\"label\":")?;
    serde_json::to_writer(&mut *writer, &entry.label)?;
    writer.write_all(b",\"eventName\":")?;
    serde_json::to_writer(&mut *writer, &entry.event_name)?;
    writer.write_all(b",\"payload\":")?;
    Ok(())
}

fn write_json_string_file(writer: &mut impl Write, path: &Path) -> AppResult<()> {
    writer.write_all(b"\"")?;
    let mut source = BufReader::new(File::open(path)?);
    let mut buffer = [0_u8; 32 * 1024];
    loop {
        let read = source.read(&mut buffer)?;
        if read == 0 {
            break;
        }
        for byte in &buffer[..read] {
            match *byte {
                b'"' => writer.write_all(br#"\""#)?,
                b'\\' => writer.write_all(br#"\\"#)?,
                b'\n' => writer.write_all(br#"\n"#)?,
                b'\r' => writer.write_all(br#"\r"#)?,
                b'\t' => writer.write_all(br#"\t"#)?,
                0x00..=0x1f => write!(writer, "\\u{:04x}", byte)?,
                byte => writer.write_all(&[byte])?,
            }
        }
    }
    writer.write_all(b"\"")?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use base64::Engine;
    use uuid::Uuid;

    use crate::services::{
        realtime_payload_service::{RealtimePayloadStore, REALTIME_INLINE_PAYLOAD_LIMIT},
        realtime_service::{
            RealtimeConnectionStatus, RealtimeSessionSnapshot, RealtimeTranscriptDirection,
            RealtimeTranscriptEntry, RealtimeTranscriptKind,
        },
    };

    use super::*;

    #[tokio::test]
    async fn large_file_backed_payload_export_streams_complete_base64() {
        let root = std::env::temp_dir().join(format!("postnot-export-test-{}", Uuid::new_v4()));
        let store = RealtimePayloadStore::new(root.clone());
        store.reset().await.expect("reset");
        let bytes = vec![0x5a; REALTIME_INLINE_PAYLOAD_LIMIT + 1];
        let payload = store.store_binary(&bytes).await.expect("store");
        assert!(matches!(payload, RealtimePayload::File { .. }));
        let snapshot = RealtimeSessionSnapshot {
            connection_id: "large-export".to_string(),
            generation: 2,
            last_sequence: 4,
            status: RealtimeConnectionStatus::Disconnected,
            status_message: "Disconnected".to_string(),
            transcript_size_bytes: bytes.len() as u64,
            transcript: vec![RealtimeTranscriptEntry {
                id: "entry".to_string(),
                connection_id: "large-export".to_string(),
                generation: 2,
                sequence: 4,
                occurred_at: "2026-01-01T00:00:00Z".to_string(),
                direction: RealtimeTranscriptDirection::Received,
                kind: RealtimeTranscriptKind::Binary,
                label: "Received binary".to_string(),
                event_name: None,
                payload: Some(payload),
            }],
        };
        let destination = root.join("export.json");
        write_transcript_export(&destination, &snapshot, &store).expect("export");
        let json: serde_json::Value =
            serde_json::from_slice(&std::fs::read(&destination).expect("read export"))
                .expect("valid JSON");
        let encoded = json["transcript"][0]["payload"]["text"]
            .as_str()
            .expect("base64 text");
        assert_eq!(BASE64.decode(encoded).expect("decode"), bytes);
        let _ = tokio::fs::remove_dir_all(root).await;
    }

    #[test]
    fn suggested_payload_names_drop_paths_and_control_characters() {
        assert_eq!(
            sanitize_suggested_file_name("../bad/name\u{0}.json"),
            "name.json"
        );
        assert_eq!(
            sanitize_suggested_file_name("payload.very-long-invalid-extension"),
            "payload-very-long-invalid-extension.bin"
        );
    }
}
